src/Service/Bulk/BulkUserCreationService.php

```diff
@@ -0,0 +1,327 @@
+        LtiInstanceRepository $ltiInstanceRepository,
+        UserPasswordHasherInterface $passwordHasher,
+        NativeUserRepository $userRepository,
+        AssignmentIngester $assignmentIngester,
+        Filesystem $filesystem,
+        CsvWriterBuilder $csvWriterBuilder,
+        string $automateUserListPath
+    ) {
+        $this->lineItemRepository = $lineItemRepository;
+        $this->assignmentRepository = $assignmentRepository;
+        $this->ltiInstanceRepository = $ltiInstanceRepository;
+        $this->passwordHasher = $passwordHasher;
+        $this->userRepository = $userRepository;
+        $this->assignmentIngester = $assignmentIngester;
+        $this->filesystem = $filesystem;
+        $this->csvWriterBuilder = $csvWriterBuilder;
+        $this->automateUserListPath = $automateUserListPath;
+    }
+
+    /**
+     * @throws LineItemNotFoundException
+     */
+    public function processData(
+        array $lineItemIds,
+        array $lineItemSlugs,
+        array $userPrefix,
+        int $batchSize,
+        string $groupPrefix
+    ): array {
+
+        $notExistLineItemsArray = $userGroupIds = [];
+
+        if (!empty($lineItemIds) || !empty($lineItemSlugs)) {
+            $this->lineItemIds = $lineItemIds;
+            $this->lineItemSlugs = $lineItemSlugs;
+            $criteria = $this->getFindLineItemCriteria();
+            $lineItemsArray = ($this->lineItemRepository->findLineItemsByCriteria($criteria))->jsonSerialize();
+            if (empty($lineItemsArray)) {
+                $exceptionMessage = $this->lineItemIds
+                    ? implode(',', $this->lineItemIds) . ' LineItem Ids not exist in the system'
+                    : implode(',', $this->lineItemSlugs) . ' LineItem Slugs not exist in the system';
+                throw new LineItemNotFoundException($exceptionMessage);
+            }
+
+            $lineItemSlugArray = $this->generateSlugData($lineItemsArray);
+            $notExistLineItemsArray = $this->lineItemIds
+                ? array_diff($lineItemIds, array_keys($lineItemSlugArray))
+                : array_diff($lineItemSlugs, array_values($lineItemSlugArray));
+        }
+
+        if (empty($lineItemIds) && empty($lineItemSlugs)) {
+            $this->getAllLineItemSlugs();
+        }
+
+        $userIncrNo = $this->getLastUserAssignedToLineItems($this->lineItemSlugs);
+
+        $userGroupAssignCount = 0;
+        if ($groupPrefix) {
+            $userGroupIds = $this->getLoadBalanceGroupID($groupPrefix);
+            $userGroupAssignCount = (int) ceil(
+                count($userPrefix) * count($this->lineItemSlugs) * $batchSize / count($userGroupIds)
+            );
+        }
+
+        $totalUsersCreated = $this->createBulkUserAssignmentData(
+            $userIncrNo,
+            $userGroupIds,
+            $userPrefix,
+            $userGroupAssignCount,
+            $groupPrefix,
+            $batchSize
+        );
+
+        return [
```

Comment:
I would avoid base solutions on arrays, maybe you can create a class like `UserCretionResult` with these properties:

- message
- status
- nonExistingLineItems

WDYT?